### PR TITLE
Add USB device number matching

### PR DESCRIPTION
**Pull request by efuss**
I needed a `device` argument for USB matching in order to independently match two identical UPSen attached to the same bus.

The changes to usb-common.[hc] and libusb.c are straightforward (sans the `DEBUG_EXACT_MATCH_DEVICE`, which I modelled after `DEBUG_EXACT_MATCH_BUS`, whatever that is used for).

The changes to the various USB drivers are mechanical. Maybe some `#define` for the regex array size (or even seven more `#defines` for the magical index values) would be useful.

I documented the `device` field only where the existing `bus` field was documented (which is not the case everywhere it's actually used).

### CODE REVIEW

File: drivers/usb-common.c
Comment (jimklimov) [inline drivers/usb-common.c:380]: To be consistent with other code around, can you please add the `upsdebugx()` call for the failure, probably similar to one you added at https://github.com/networkupstools/nut/pull/974/files#diff-6aea68febe9f0fc9ad47c0d821a43073c65461e02ca9f0648ba78981482f98b8R122 ?

Otherwise, looks good to me.

**Comment by efuss** [comment]
> To be consistent with other code around
Sure. I thought I'd done that, but probably I missed that part when adapting my patch (against the stable 2.7.4 version) to Git master.

cd33fc706ff0766ac2a3cc813f7fbff7d1229d9d.

**Review by jimklimov — APPROVED** [review]
LGTM, so going to merge soon unless some other maintainers voice objections?

One minor nit, inherited from earlier blocks in the readme's, is the Linux-specific path names and `lsusb`, but this can be addressed later to disclaim those examples: `e.g. lsusb(8) on Linux` or some such wording.

**Comment by clepple** [comment]
> LGTM, so going to merge soon unless some other maintainers voice objections?
> 
> One minor nit, inherited from earlier blocks in the readme's, is the Linux-specific path names and `lsusb`, but this can be addressed later to disclaim those examples: `e.g. lsusb(8) on Linux` or some such wording.

`lsusb` is one of the first things I install from the ports tree on *BSD :-) so I don't think we need to call it out as Linux-specific. For the other instances of the device, we could maybe reference driver debug messages.

**Review by clepple — COMMENTED** [review]
I'm on the fence about this. The mechanics of the patch are solid, but device numbers are not likely to be stable. I'd maybe consider this an interim fix until someone has time to look at the physical port location stuff that is specific to libusb-1.0 (and document the options as such).

I am also concerned about how this will affect mergeability of #300.

**Comment by efuss** [comment]
> device numbers are not likely to be stable
But it's all we've got, no?

For my Masterguard UPSen, I check the "slave address" (ups.id) to make sure I'm not talking to the wrong one.

Unfortunately, the underlying logic in NUT is that once the USB matching has succeeded, the driver/protocol layer can only claim or refuse to claim, but refusing to claim will not allow another driver instance to claim the same device.

So combining device and slave_address parameter at least gives me a chance that it will attach correctly and fail if device numbers changed.

**Comment by clepple** [comment]
> > device numbers are not likely to be stable
> But it's all we've got, no?

this is the physical port location stuff I was thinking of: https://stackoverflow.com/questions/14279796/linux-libusb-get-usb-device-path

> For my Masterguard UPSen, I check the "slave address" (ups.id) to make sure I'm not talking to the wrong one. Unfortunately, the underlying logic in NUT is that once the USB matching has succeeded, the driver/protocol layer can only claim or refuse to claim, but refusing to claim will not allow another driver instance to claim the same device. So combining device and slave_address parameter at least gives me a chance that it will attach correctly and fail if device numbers changed.

I just think that if we are merging this into the main branch for a release, we should set expectations that some manual intervention will be needed when using bus and device number matching. Fixing the claim logic or testing the libusb-1.0 branch would be even better, IMHO.

**Comment by efuss** [comment]
> we should set expectations that some manual intervention will be needed when using bus and device number matching.
What about the disclaimers I just added in 92672be3?

**Comment by clepple** [comment]
Maybe "Note that device numbers are not likely to be stable across reboots or device re-plugging."

**Comment by efuss** [comment]
> Maybe "Note that device numbers are not likely to be stable across reboots or device re-plugging."
I was just in the process of re-phrasing it, but then re-consindered it.
I don't want to start nit-picking, but if you re-boot without fiddling with the cabling in betweeen, one would expect the attachments being done the same way as last time, no?

**Comment by clepple** [comment]
> I was just in the process of re-phrasing it, but then re-consindered it. I don't want to start nit-picking, but if you re-boot without fiddling with the cabling in betweeen, one would expect the attachments being done the same way as last time, no?

Ideally, yes. However, let's say you plug in, unplug, plug in, configure with device number, reboot - the device number will likely be lower after the reboot. A newer kernel might also reorder the devices.

**Comment by jimklimov** [comment]
Am I correct assuming we are still discussing about, or waiting for, a commit with better caveats-emptor warning to the documentation for this feature?

**Comment by efuss** [comment]
> Am I correct assuming we are still discussing about, or waiting for, a commit with better caveats-emptor warning to the documentation for this feature?
No idea. I regard my warning as sufficient, but I'm not the project and not in a position to make a decision.

**Comment by eski-dev** [comment]
Forgive me as I am not well versed in github etiquette, not do I have anything particularly useful to add to the conversation, however this is a feature that would be key to my current roadblock implementing two identical Tripp Lite SMART1000RM1U UPSes. Is there an estimated timeline for having this change formally added to the code?

**Comment by jimklimov** [comment]
Sorry for stalling this a bit, hope to merge soon now. Thanks for the contribution!

Just in case: was this tested for setups without a specified `device` configuration option, that this would not preclude identifying a single attached device, etc.? By code looks like it would behave same as unspecified any other such parameter, but got me wondering how we really handle empty/null values there...

**Comment by efuss** [comment]
> Just in case: was this tested for setups without a specified `device` configuration option
I'm afraid not. I always have two identical UPSen connected to a server.

**Pull request merged**
